igcollect/sfp_dom.py

```diff
@@ -0,0 +1,371 @@
+                from exc
+        else:
+            raise
+
+    ee = ethtool_eeprom()
+    ee.cmd = ETHTOOL_GMODULEEEPROM
+    ee.len = em.eeprom_len
+    ee.offset = 0
+    ifr.ifr_data.ethtool_eeprom_ptr = ctypes.pointer(ee)
+
+    fcntl.ioctl(sock, SIOCETHTOOL, ifr)
+
+    if em.type in (
+            ETH_MODULE_SFF_8079, ETH_MODULE_SFF_8636, ETH_MODULE_SFF_8436
+    ):
+        raise UnsupportedSFPException(f'SFP type {em.type} is not supported')
+    elif em.type == ETH_MODULE_SFF_8472:
+        return SFPData(ee.data, interface)
+
+    return None
+
+
+class ethtool_modinfo(ctypes.Structure):
+    _fields_ = [
+        ('cmd', ctypes.c_uint32),
+        ('type', ctypes.c_uint32),
+        ('eeprom_len', ctypes.c_uint32),
+        ('reserved', ctypes.c_uint32 * 8)
+    ]
+
+
+class ethtool_eeprom(ctypes.Structure):
+    _fields_ = [
+        ('cmd', ctypes.c_uint32),
+        ('magic', ctypes.c_uint32),
+        ('offset', ctypes.c_uint32),
+        ('len', ctypes.c_uint32),
+        ('data', ctypes.c_byte * 512)
+    ]
+
+
+class ifr_data(ctypes.Union):
+    _fields_ = [
+        ('ethtool_modinfo_ptr', ctypes.POINTER(ethtool_modinfo)),
+        ('ethtool_eeprom_ptr', ctypes.POINTER(ethtool_eeprom)),
+    ]
+
+
+class ifreq(ctypes.Structure):
+    _fields_ = [
+        ('ifr_name', (ctypes.c_ubyte * IFNAMSIZ)),
+        ('ifr_data', ifr_data)
+    ]
+
+
+class SFPData:
+    def __init__(self, data, interface):
+        self._data = data
+        self.interface = interface
+        self.metrics = {
+            'temperature': {
+                'current': None,
+                'low_warning': None,
+                'high_warning': None,
+                'low_critical': None,
+                'high_critical': None,
+            },
+            'bias_current': {
+                'current': None,
+                'low_warning': None,
+                'high_warning': None,
+                'low_critical': None,
+                'high_critical': None,
+            },
+            'sfp_voltage': {
+                'current': None,
+                'low_warning': None,
+                'high_warning': None,
+                'low_critical': None,
+                'high_critical': None,
+            },
+            'tx_power': {
+                'current': None,
+                'low_warning': None,
+                'high_warning': None,
+                'low_critical': None,
+                'high_critical': None,
+            },
+            'rx_power': {
+                'current': None,
+                'low_warning': None,
+                'high_warning': None,
+                'low_critical': None,
+                'high_critical': None,
+            },
+        }
+        self._parse_eeprom_data()
+
+    def _parse_eeprom_data(self):
+        # Check if SFP data is externally calibrated
+        self.external_calibration = self._data[SFF_A0_DOM] & SFF_A0_DOM_EXTCAL
```

Comment:
You are never using that again. So, I doubt that it is needed to store it on `self`.

Besides, I would also suggest to do that directly during `__init__` or can `self._data` change between `__init__` and `_parse_eeprom_data` execution?

Reply:
It makes sense. 
I moved the check up and removed the `self`.